p2p/lightchain.py

```diff
@@ -245,35 +263,105 @@ def peer_block_height(peer: LESPeer) -> int:
         await self.wait_until_finished()
         self.logger.debug("LightPeerChain finished")
 
+    async def _wait_for_reply(self, request_id: int) -> Dict[str, Any]:
+        reply = None
+        got_reply = asyncio.Event()
+
+        def callback(r):
+            nonlocal reply
+            reply = r
+            got_reply.set()
+
+        self._pending_replies[request_id] = callback
+        await wait_with_token(got_reply.wait(), token=self.cancel_token, timeout=self.reply_timeout)
+        return reply
+
     @alru_cache(maxsize=1024, cache_exceptions=False)
-    async def get_block_header_by_hash(self, block_hash: Hash32) -> BaseBlock:
+    async def get_block_header_by_hash(self, block_hash: Hash32) -> BlockHeader:
         peer = await self.get_best_peer()
-        self.logger.debug("Fetching header %s from %s", encode_hex(block_hash), peer)
-        return await peer.get_block_header_by_hash(block_hash, self.cancel_token)
+        return await self._get_block_header_by_hash(peer, block_hash)
 
     @alru_cache(maxsize=1024, cache_exceptions=False)
     async def get_block_body_by_hash(self, block_hash: Hash32) -> BaseBlock:
         peer = await self.get_best_peer()
         self.logger.debug("Fetching block %s from %s", encode_hex(block_hash), peer)
-        return await peer.get_block_by_hash(block_hash, self.cancel_token)
+        request_id = gen_request_id()
```

Comment:
Also, I was surprised to see how `gen_request_id` is implemented (I sort of expected it to return a `uuid.uuid4()` or something.  Is there a reason for it's current implementation?

Reply:
Nope. I've changed it to return `big_endian_to_int(uuid4().bytes)`

Reply:
Actually, `uuid4().int`

Reply:
And it turns out a uuid4 is too long (16 bytes) to use in the request_id field. I'd need to use only 8 bytes of it, so I guess I'll stick with os.urandom() for now, unless you have another suggestion?

Reply:
No other suggestions 👍